### PR TITLE
CNI: Install a global NetworkAttachmentDefinition

### PR DESCRIPTION
**Pull request by jwendell**
This simplifies installation and operation on OpenShift clusters. By having a global NetworkAttachmentDefinition installed on the `default` namespace users don't need to install them in each application namespace.

The pod injection - in the openshift profile - needs to point to the global resource in the `default` namespace for this to work.

With this change we can remove the [last section in the OpenShift Setup page](https://istio.io/v1.16/docs/setup/platform-setup/openshift/#additional-requirements-for-the-application-namespace).

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
LGTM personally. I don't have much expertise in this and I don't have permissions to approve either, so will leave for @jacob-delgado who I think has some openshift expertise + permission

**Review by costinm — COMMENTED** [review]
Looks good.  Few 'bike sheding' comments:

- if I understand correctly, namespace can be specified - it doesn't have to be 'default' ? 
If that's the case, wouldn't istio-system or kube-system be the right value ? The 'default'
namespace is usually used for examples.

- would it make sense to instead add a "cni.provider" - with values like cillium, multus, etc ? 
- or 'cni.platform' - with 'openshift', 'gke-calico', 'gke-dpv2', 'azure', etc ? 

Few users will understand 'networkAttachmentDefinition' in the config.

**Review by dgn — APPROVED** [review]
lgtm to in general. not a fan of the long config names, but... :man_shrugging: wouldn't it be enough to check for `providers.multus.enabled`? Disabling the global NetworkAttachmentDefinition has no benefits from my perspective

File: manifests/charts/istio-cni/values.yaml [hunk 1]
```diff
@@ -44,6 +44,20 @@ cni:
   # Allow the istio-cni container to run in privileged mode, needed for some platforms (e.g. OpenShift)
   privileged: false
 
+  providers:
```
Comment (costinm) [inline manifests/charts/istio-cni/values.yaml:47]: I'm ok with this - but I was thinking of a simpler 

# Upstream CNI provider if known
# TODO: autodetect if possible with helm
provider: "default"

And if provider == multus, just set that thing on. Is there any reason to NOT set it when multus is used,
or to use a different namespace ? If not - no need to have another install option.

File: manifests/charts/istio-cni/values.yaml [hunk 2]
```diff
@@ -44,6 +44,20 @@ cni:
...
+      globalNetworkAttachmentDefinition:
+        # If enabled, a global NetworkAttachmentDefinition will be created in the namespace below. Note that the cluster
+        # needs to have this CRD (network-attachment-definitions.k8s.cni.cncf.io) installed beforehand.
+        enabled: false
+
+        # The namespace the NetworkAttachmentDefinition global resource will be created on. This resource will be referenced
+        # by all pods that have a sidecar. This is done by putting the annotation `k8s.v1.cni.cncf.io/networks: <ns>/istio-cni` into the pod.
+        # By default `default` namespace is allowed to be referenced by all namespaces.
+        # For more information on this, see https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/configuration.md#namespace-isolation
+        namespace: default
```
Comment (costinm) [inline manifests/charts/istio-cni/values.yaml:59]: If we don't expect users to use a different one - I would just remove it as an option.

File: manifests/profiles/openshift.yaml [hunk 3]
```diff
@@ -16,6 +16,11 @@ spec:
...
       privileged: true
+      providers:
+        multus:
+          globalNetworkAttachmentDefinition:
+            enabled: true
+            namespace: default
     sidecarInjectorWebhook:
       injectedAnnotations:
-        k8s.v1.cni.cncf.io/networks: istio-cni
+        k8s.v1.cni.cncf.io/networks: default/istio-cni
```
Comment (costinm) [inline manifests/profiles/openshift.yaml:26]: Don't have enough context - is multus an openshift-only thing ?

**Review by costinm — APPROVED** [review]
Looks good - left cosmetic comments, feel free to ignore.

**Comment by jwendell** [comment]
@costinm @howardjohn I made the suggested changes to make it simpler. If you are ok with them, can you remove the DNM label?

File: manifests/profiles/openshift.yaml (hunk 3 above)
Comment (jacob-delgado) [reply manifests/profiles/openshift.yaml:22]: No, it's just another CNI. It's used to create separate vlans for things like the k8s control plane and the k8s dataplane. You can create more vlans in addition.

**Review by jacob-delgado — APPROVED** [review]
Great change. Helping users manage openshift would be great.

**Comment by jacob-delgado** [comment]
@jwendell you think we can get this merged into 1.18?

**Comment by jwendell** [comment]
> @jwendell you think we can get this merged into 1.18?

I have no objections, although there's another change that I'll submit soon to get rid of the anyuid requirements on openshift. Those two together will eliminate the need of having the openshift install doc page <3

**Review by howardjohn — APPROVED** [review]

**Comment by jacob-delgado** [comment]
> > @jwendell you think we can get this merged into 1.18?
> 
> I have no objections, although there's another change that I'll submit soon to get rid of the anyuid requirements on openshift. Those two together will eliminate the need of having the openshift install doc page <3

If we could get both of these in for 1.18 that would be awesome. Many of our users want to adopt OpenShift, but when they first start out they struggle getting some of these minor changes in.

**Pull request merged**